sop-core/src/Data/SOP/NS.hs

```diff
@@ -146,13 +163,45 @@ import Data.SOP.Sing
 -- > S (Z (I True)) :: NS I       '[ Char, Bool ]
 -- > S (Z (K 1))    :: NS (K Int) '[ Char, Bool ]
 --
-data NS :: (k -> Type) -> [k] -> Type where
-  Z :: f x -> NS f (x ': xs)
-  S :: NS f xs -> NS f (x ': xs)
+data NS (f :: k -> *) (xs :: [k]) = NS !Int Any
+
+-- | View on NP
+--
+-- This is only used internally, for the definition of the pattern synonyms.
+data ViewNS (f :: k -> *) (xs :: [k]) where
+  IsZ :: f x     -> ViewNS f (x ': xs)
+  IsS :: NS f xs -> ViewNS f (x ': xs)
+
+viewNS :: NS f xs -> ViewNS f xs
+viewNS (NS i x)
+  | i == 0    = unsafeCoerce (IsZ (unsafeCoerce x))
+  | otherwise = unsafeCoerce (IsS (NS (i - 1) x))
+
+pattern Z :: forall f xs' . ()
+          => forall x xs . (xs' ~ (x ': xs)) => f x -> NS f xs'
+pattern Z x <- (viewNS -> IsZ x)
+  where
+    Z x = NS 0 (unsafeCoerce x)
+
+pattern S :: forall f xs' . ()
+          => forall x xs . (xs' ~ (x ': xs)) => NS f xs -> NS f xs'
+pattern S p <- (viewNS -> IsS p)
+  where
+    S (NS i x) = NS (i + 1) x
+
+#if __GLASGOW_HASKELL__ >= 802
+{-# COMPLETE Z, S #-}
+#endif
 
-deriving instance All (Show `Compose` f) xs => Show (NS f xs)
-deriving instance All (Eq   `Compose` f) xs => Eq   (NS f xs)
-deriving instance (All (Eq `Compose` f) xs, All (Ord `Compose` f) xs) => Ord (NS f xs)
+instance All (Show `Compose` f) xs => Show (NS f xs) where
+  show ns @ (NS i _) =
```

Comment:
Add TODO to implement this properly. To be backward compatible this needs to print `S`s and `Z`s.

BTW. The space between `@` and `(NS ...` will break on GHC-9.0 due https://github.com/ghc-proposals/ghc-proposals/pull/229

```
src/Data/SOP/NS.hs:197:3: error:
    Parse error in pattern: show
    In a function binding for the ‘@’ operator.
    Perhaps you meant an as-pattern, which must not be surrounded by whitespace
    |
197 |   show ns @ (NS i _) =
```

(Otherwise this patch is GHC-9.0 friendly)